connectors/qualys/connector/data_handler.py

```diff
@@ -0,0 +1,287 @@
+                    key = interface['HostAssetInterface'].get('macAddress', "")
+                    if not key:
+                        continue
+                    res['_key'] = key
+                    res['interface'] = interface['HostAssetInterface'].get('interfaceName', "")
+
+                    self.add_collection('macaddress', res, '_key')
+
+                    # asset mac address edge creation
+                    asset_macaddress = {'_from_external_id': str(obj['HostAsset']['id']),
+                                        '_to': "macaddress/" + interface['HostAssetInterface'].get('macAddress',
+                                                                                                   "")}
+                    self.add_edge('asset_macaddress', asset_macaddress)
+
+                    # ip address mac address edge creation
+                    ipaddress_mac = {'_from': "ipaddress/" + interface['HostAssetInterface']['address'],
+                                     '_to': "macaddress/" + interface['HostAssetInterface'].get('macAddress', "")}
+                    self.add_edge('ipaddress_macaddress', ipaddress_mac)
+
+    # Create hostname Object as per CAR data model from data source
+    def handle_hostname(self, obj):
+        if obj:
+            res = self.copy_fields(obj, )
+            res['host_name'] = obj['HostAsset']['name'].lower()
+            key = obj['HostAsset'].get('dnsHostName', "").lower()
+
+            if key:
+                res['_key'] = key
+            else:
+                res['_key'] = res['host_name']
+
+            self.add_collection('hostname', res, '_key')
+
+            # asset hostname edge creation
+            asset_hostname = {'_from_external_id': str(obj['HostAsset']['id']), '_to': 'hostname/' + res['_key']}
+            self.add_edge('asset_hostname', asset_hostname)
+
+    # Create application Object as per CAR data model from data source
+    def handle_application(self, obj):
+        if obj and 'vmdrVulnList' in obj['HostAsset'] and 'applications' in obj['HostAsset']:
+            for app in obj['HostAsset']['applications']:
+                for vuln in obj['HostAsset']['vmdrVulnList']:
+                    if app['productName'].lower() in vuln['RESULTS'].lower():
+                        res = self.copy_fields(obj, )
+                        res['name'] = app['fullName']
+                        res['external_id'] = str(app['id'])
+                        self.add_collection('application', res, 'external_id')
+
+                        # asset application edge creation
+                        asset_application = {'_from_external_id': str(obj['HostAsset']['id']),
+                                             '_to_external_id': str(app['id'])}
+                        self.add_edge('asset_application', asset_application)
+
+                        # application vulnerability edge creation
+                        application_vulnerability = {'_from_external_id': str(app['id']),

```

Comment:
Are these vulnerability specific to this application in the asset? If it is just same vuln we have for asset then we dont need edge between application_vulnerability.

Reply:
Yes, these vulnerabilities are specific to this application in the asset.

Our assumption now is that if vulnerable application is present in asset,  we can map asset to vulnerability(asset_vulnerability). 'application_vulnerability' edge is not required and can be removed.

Can we retain application node?